File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/validator/ValidationServiceTest.java
package cz.fb.manaus.reactor.betting.validator;

import cz.fb.manaus.core.model.Bet;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.reactor.betting.BetContext;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.Optional;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ValidationServiceTest extends AbstractLocalTestCase {
    @Autowired
    private ValidationService service;

    @Test
    public void testReduceAcceptReject() {
        assertThat(service.reduce(List.of(ValidationResult.ACCEPT, ValidationResult.REJECT, ValidationResult.ACCEPT)), is(ValidationResult.REJECT));
        assertThat(service.reduce(List.of(ValidationResult.ACCEPT, ValidationResult.ACCEPT)), is(ValidationResult.ACCEPT));
    }

    @Test
    public void testDowngradeAccepting() {
        assertThat(new TestValidator(ValidationResult.REJECT).isDowngradeAccepting(), is(true));
    }

    @Test
    public void testDowngradePrice() {
        checkDownGrade(2d, ValidationResult.ACCEPT);
    }

    @Test
    public void testUpgradePrice() {
        checkDownGrade(2.4d, null);
    }

    private void checkDownGrade(double newPrice, ValidationResult expected) {
        var oldBet = mock(Bet.class);
        when(oldBet.getRequestedPrice()).thenReturn(new Price(2.2d, 2d, Side.LAY));
        var rejecting = new TestValidator(ValidationResult.REJECT);
        var result = service.handleDowngrade(
                Optional.of(new Price(newPrice, 2d, Side.LAY)),
                Optional.of(oldBet), rejecting);
        assertThat(result.orElse(null), is(expected));
        assertThat(rejecting.validate(null), is(ValidationResult.REJECT));
    }


    private static class TestValidator implements Validator {

        private final ValidationResult result;

        private TestValidator(ValidationResult result) {
            this.result = result;
        }

        @Override
        public ValidationResult validate(BetContext context) {
            return result;
        }
    }
}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/validator/ValidationServiceTest.kt
package cz.fb.manaus.reactor.betting.validator

import cz.fb.manaus.core.model.Bet
import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.core.test.AbstractTestCase
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.HOME_EVENT_BACK
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ValidationServiceTest : AbstractTestCase() {
    @Autowired
    private lateinit var service: ValidationService

    @Test
    fun `reduce OK - NOP - DROP combinations`() {
        val results = listOf(
                ValidationResult.OK,
                ValidationResult.NOP,
                ValidationResult.DROP,
                ValidationResult.OK
        )
        assertEquals(
                ValidationResult.DROP,
                service.reduce(results)
        )
        assertEquals(ValidationResult.OK, service.reduce(results.filter { it == ValidationResult.OK }))
        assertEquals(ValidationResult.NOP, service.reduce(results.filter { it != ValidationResult.DROP }))
    }

    @Test
    fun `downgrade accepting by default`() {
        assertTrue(MockValidator(ValidationResult.DROP).isDowngradeAccepting)
    }

    @Test
    fun `downgrade price`() {
        checkDownGrade(2.0, true)
    }

    @Test
    fun `upgrade price`() {
        checkDownGrade(2.4, false)
    }

    private fun checkDownGrade(newPrice: Double, expected: Boolean) {
        val oldBet = Bet(
                marketId = "1", selectionId = 1, requestedPrice = Price(2.2, 2.0, Side.LAY),
                placedDate = Instant.now(), matchedAmount = 5.0
        )
        val dropping = MockValidator(ValidationResult.DROP)
        val result = service.isDowngrade(Price(newPrice, 2.0, Side.LAY), oldBet)
        assertEquals(expected, result)
        assertEquals(ValidationResult.DROP, dropping.validate(HOME_EVENT_BACK))
    }


    private data class MockValidator(private val result: ValidationResult) : Validator {

        override fun validate(event: BetEvent): ValidationResult {
            return result
        }

    }
}
